include/oneapi/dpl/internal/function.h

```diff
@@ -61,32 +61,33 @@ using oneapi::dpl::__par_backend_hetero::__internal::is_hetero_iterator;
 #if _ONEDPL_BACKEND_SYCL
 // Helpers used to get indexable access to the data passed to the SYCL implementation of an
 // algorithm from either a SYCL iterator or a USM pointer.
-template <cl::sycl::access::mode Mode, typename Iterator>
+template <cl::sycl::access::mode Mode, typename Policy, typename Iterator>
 auto
-get_access(Iterator i, typename ::std::enable_if<is_hetero_iterator<Iterator>::value, void>::type* = nullptr)
+get_access(Policy, Iterator i, typename ::std::enable_if<is_hetero_iterator<Iterator>::value, void>::type* = nullptr)
     -> decltype(i.get_buffer().template get_access<Mode>())
 {
     return i.get_buffer().template get_access<Mode>();
 }
 
-template <cl::sycl::access::mode Mode, typename Iterator>
+template <cl::sycl::access::mode Mode, typename Policy, typename Iterator>
 Iterator
-get_access(Iterator i, typename ::std::enable_if<!is_hetero_iterator<Iterator>::value, void>::type* = nullptr)
+get_access(Policy, Iterator i, typename ::std::enable_if<!is_hetero_iterator<Iterator>::value, void>::type* = nullptr)
 {
     return i;
 }
 
-template <cl::sycl::access::mode Mode, typename T>
+template <cl::sycl::access::mode Mode, typename Policy, typename T>
 counting_iterator<T>
-get_access(counting_iterator<T> i)
+get_access(Policy, counting_iterator<T> i)
 {
     return i;
 }
 
-template <cl::sycl::access::mode Mode, typename T>
+template <cl::sycl::access::mode Mode, typename Policy, typename T>
 T*
-get_access(T* ptr)
+get_access(Policy policy, T* ptr)
 {
+    assert(sycl::get_pointer_type(ptr, policy.queue().get_context()) == sycl::usm::alloc::shared);
```

Comment:
Why do we need to extend by an extra argument for policy? Looks like this is the only place it's used.

Reply:
Because we  have several overloads of internal::get_access with the same call pattern.
As alternative way we would use a "context"...  But I think usage of "policy" is more general way.